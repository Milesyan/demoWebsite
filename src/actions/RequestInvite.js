import ApiClient from '../network/ApiClient';
import { ActionTypes } from './ActionTypes';
import { ServerError } from '../network/errors';

export function requestInviteLoading() {
  return {
    type: ActionTypes.REQUEST_INVITE_LOADING,
  };
}

export function requestInviteSuccess() {
  return {
    type: ActionTypes.REQUEST_INVITE_SUCCESS,
  };
}

export function requestInviteFail(msg) {
  return {
    type: ActionTypes.REQUEST_INVITE_FAIL,
    errorMsg: msg
  };
}

export function resetRequestInviteDialog() {
  return {
    type: ActionTypes.RESET_DIALOG_STATUS
  }
}

export function requestInvite(fullName, email) {
  return async (dispatch) => {
    dispatch(requestInviteLoading());
    try {
      const r = await ApiClient.requestInvite(fullName, email)
      if (r === 'Registered') {
        dispatch(requestInviteSuccess())
      } else {
        throw new Error(`Unknown 200 response >>> ${r}`);
      }
    } catch (error) { 
      if (error instanceof ServerError) {
        dispatch(requestInviteFail(error.message.errorMessage || "Server Error with mo messsage"))
      } else {
        dispatch(requestInviteFail("Unkown Error"));
        throw error;
      }
    }
  };
}
