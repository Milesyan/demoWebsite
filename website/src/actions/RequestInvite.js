import ApiClient from '../network/ApiClient';
import { ActionTypes } from './ActionTypes';
import { HttpError, ServerError } from '../network/errors';

function requestInviteLoading() {
  return {
    type: ActionTypes.REQUEST_INVITE_LOADING,
  };
}

function requestInviteSuccess() {
  return {
    type: ActionTypes.REQUEST_INVITE_SUCCESS,
  };
}

function requestInviteFail(msg) {
  return {
    type: ActionTypes.REQUEST_INVITE_FAIL,
    errorMsg: msg
  };
}

export function resetRequestInviteDialog() {
  return {
    type: ActionTypes.RESET_DIALOG_STATUS
  }
}

export function requestInvite(fullName, email) {
  return async (dispatch) => {
    dispatch(requestInviteLoading());
    try {
      const r = await ApiClient.requestInvite(fullName, email)
      if (r === 'Registered') {
        dispatch(requestInviteSuccess())
      } else {
        throw new Error(`Unknown 200 response >>> ${r}`);
      }
    } catch (error) { 
      if (error instanceof ServerError) {
        dispatch(requestInviteFail(error.message))
      } else {
        dispatch(requestInviteFail("Unkown Error"));
        throw error;
      }
    }
  };
}
