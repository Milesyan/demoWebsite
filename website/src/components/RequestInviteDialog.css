.request-dialog {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 30px;
}

.request-dialog-title {
  font-size: 24px;
  margin-bottom: 40px;
}
.request-dialog-input {
  width: 80%;
  padding: 12px;
  margin-bottom: 12px;
}
