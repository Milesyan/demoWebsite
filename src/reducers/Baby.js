import { ActionTypes } from '../actions/ActionTypes';
import { baby } from './__mock__/baby';

const initialState = {
  baby
}

export default (state = initialState, { type, payload }) => {
  switch (type) {

  case ActionTypes.SET_BABY_INFO:
    return { ...state, ...payload }

  default:
    return state
  }
}
