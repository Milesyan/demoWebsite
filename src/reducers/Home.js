import { ActionTypes } from '../actions/ActionTypes';
export const HOME_STATUS = {
  initial: Symbol(),
  preview: Symbol(),
  process: Symbol(),
  done: Symbol()
}
const defaultState = {
  status: HOME_STATUS.initial
}
const Home = (state = defaultState, action) => {
  switch (action.type) {
    case ActionTypes.SET_HOME_STATUS_INITIAL:
      return { status: HOME_STATUS.initial };
    case ActionTypes.SET_HOME_STATUS_PREVIEW:
      return { status: HOME_STATUS.preview };
    case ActionTypes.SET_HOME_STATUS_PROCESS:
      return { status: HOME_STATUS.process };
    case ActionTypes.SET_HOME_STATUS_DONE:
      return { status: HOME_STATUS.done };
    default:
      return state;
  }
}

export default Home;