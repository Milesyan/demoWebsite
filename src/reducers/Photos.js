import { ActionTypes } from '../actions/ActionTypes';
import { photoPosts, photos } from './__mock__/photos';
import produce from "immer"


const defaultState = {
  photoPosts,
  photoEntities: photos
}
const Home = (state = defaultState, action) => {
  switch (action.type) {
    case ActionTypes.UPDATE_IMAGE_INFO:
      const { id, width, height} = action.payload;
      const nextState = produce(state, draftState => {
        let photo = draftState.photoEntities[id]
        photo.width = width;
        photo.height = height;
      })
      return nextState;
    case ActionTypes.SET_POSTS:
      const postState = produce(state, draft => {
        draft.photoPosts = action.payload
      })
      return postState;
    case ActionTypes.SET_PHOTOS:
      const photosState = produce(state, draft => {
        draft.photoEntities = action.payload
      })
      return photosState;
    default:
      return state;
  }
}

export default Home;