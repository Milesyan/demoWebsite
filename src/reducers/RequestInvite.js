import { ActionTypes } from '../actions/ActionTypes';

export const DIALOG_STATE = {
  initial: -1,
  loading: 0,
  success: 1,
  error: 2
}
const defaultState = {
  status: DIALOG_STATE.initial, 
  message: null
}
const RequestInvite = (state = defaultState, action) => {
  switch (action.type) {
    case ActionTypes.REQUEST_INVITE_LOADING:
      return { status: DIALOG_STATE.loading, message: null };
    case ActionTypes.REQUEST_INVITE_SUCCESS:
      return { status: DIALOG_STATE.success, message: null };
    case ActionTypes.REQUEST_INVITE_FAIL:
      return { status: DIALOG_STATE.error, message: action.errorMsg }
    case ActionTypes.RESET_DIALOG_STATUS:
      return { status: DIALOG_STATE.initial, message: null };
    default:
      return state;
  }
}

export default RequestInvite;