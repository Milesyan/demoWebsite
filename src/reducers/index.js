import { combineReducers } from 'redux';
import RequestInvite from './RequestInvite';
export default combineReducers({
  RequestInvite
})
