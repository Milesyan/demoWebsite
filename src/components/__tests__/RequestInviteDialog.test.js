/* eslint-disable no-undef */ /*eslint-disable react/react-in-jsx-scope*/

import { RequestInviteDialog } from '../RequestInviteDialog';
import { DIALOG_STATE } from '../../reducers/RequestInvite';

const setup = overrideProps => {
  const initProps = {
    dialogStatus: DIALOG_STATE.initial,
    dialogMessage: '',
    onClose: jest.fn(),
    resetRequestInviteDialog: jest.fn(),
    requestInvite: jest.fn()
  };
  const props = { ...initProps, ...overrideProps};
  const wrapper = shallow(
    <RequestInviteDialog {...props}/>
  );
  return { wrapper, props };
}

describe('RequestInviteDialog snapshots', () => {

  it('should match snapshot in inital status', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot in loading status', () => {
    const { wrapper } = setup({dialogStatus: DIALOG_STATE.loading});
    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot in success status', () => {
    const { wrapper } = setup({dialogStatus: DIALOG_STATE.success});
    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot in error status with dialogMessage', () => {
    const { wrapper } = setup({dialogStatus: DIALOG_STATE.error, dialogMessage: "EXAMPLE ERROR MESSAGE"});
    expect(wrapper).toMatchSnapshot();
  });


});

describe('RequestInviteDialog functions', () => {

    it('should be able to close modal if modal is not in loading status', () => {
      const { wrapper, props } = setup();
      wrapper.simulate('click');
      expect(props.onClose).toBeCalled();

      wrapper.setProps({dialogStatus: DIALOG_STATE.success});
      wrapper.simulate('click');
      expect(props.onClose).toBeCalledTimes(2);
    
      wrapper.setProps({dialogStatus: DIALOG_STATE.error});
      wrapper.simulate('click');
      expect(props.onClose).toBeCalledTimes(3);
    });

    it('should not be able to close modal if the modal is in loading status', () => {
      const mockClose = jest.fn();
      const { wrapper }= setup({dialogStatus: DIALOG_STATE.loading, onClose: mockClose});
      wrapper.simulate('click');
      expect(mockClose).not.toBeCalled();
    });

    it('should have a disabled send button as default, and an active button when all inputs are valid', () => {
      const { wrapper } = setup();
      const sendBtn =  wrapper.findWhere(n => n.props().children === 'Send');
      expect(sendBtn.props().isDisabled).toBeTruthy();
      wrapper.setState({fullName: "EXAMPLE NAME", email: "test@ttt.com", confirmedEmail: "test@ttt.com"});
      wrapper.update();
      const sendBtnUpdated = wrapper.findWhere(n => n.props().children === 'Send');
      expect(sendBtnUpdated.props().isDisabled).toBeFalsy();
      expect(wrapper).toMatchSnapshot();
    });

    it('should send action to redux when onSend is triggered', () => {
      const { wrapper, props } = setup();
      wrapper.setState({fullName: "TEST ACTIOn", email: "test@action.com", confirmedEmail: "test@action.com"});
      wrapper.findWhere(n => n.props().children === 'Send').simulate('click');
      expect(props.requestInvite).toHaveBeenCalledWith("TEST ACTIOn", "test@action.com");
    });

    it('should call reset dialog action when component is unmounted', () => {
      const { wrapper, props } = setup();
      wrapper.instance().componentWillUnmount();
      expect(props.resetRequestInviteDialog).toBeCalled();
    });

});

describe('RequestInviteDialog input onChange validates', () => {
  
  it('should set correct error when onNameChanged is called', () => {
    const { wrapper } = setup();
    const instance = wrapper.instance()
    instance.onNameChanged({target: {value: 'ab'}});
    expect(wrapper.state()).toMatchSnapshot("Error state for short name");
    instance.onNameChanged({target: {value: ''}});
    expect(wrapper.state()).toMatchSnapshot("Error state for empty name");
    instance.onNameChanged({target: {value: 'Good Name'}});
    expect(wrapper.state()).toMatchSnapshot("Error state for valid name");
  });
  
  it('should set correct error when onEmailChanged is called', () => {
    const { wrapper } = setup();
    const instance = wrapper.instance()
    instance.onEmailChanged({target: {value: ''}});
    expect(wrapper.state()).toMatchSnapshot("Error state for empty email");
    instance.onEmailChanged({target: {value: 'invalid'}});
    expect(wrapper.state()).toMatchSnapshot("Error state for invalid email");
    instance.onEmailChanged({target: {value: 'test@valid.com'}});
    expect(wrapper.state()).toMatchSnapshot("Error state for valid email");
  });
  
  it('should set correct error when onConfirmedEmailChanged is called', () => {
    const { wrapper } = setup();
    const instance = wrapper.instance()
    wrapper.setState({
      email: "miles@ttt.com"
    })
    instance.onConfirmedEmailChanged({target: {value: ''}});
    expect(wrapper.state()).toMatchSnapshot("Error state for empty confirmed email");
    instance.onConfirmedEmailChanged({target: {value: 'invalidCOnfimed'}});
    expect(wrapper.state()).toMatchSnapshot("Error state for invalid email");
    instance.onConfirmedEmailChanged({target: {value: 'test@valid.com'}});
    expect(wrapper.state()).toMatchSnapshot("Error state for not matched confirmed email");
    instance.onConfirmedEmailChanged({target: {value: 'miles@ttt.com'}});
    expect(wrapper.state()).toMatchSnapshot("Error state for matched confirmed email");
  });
  
});
