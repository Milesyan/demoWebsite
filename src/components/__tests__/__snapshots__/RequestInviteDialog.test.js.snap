// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestInviteDialog functions should have a disabled send button as default, and an active button when all inputs are valid 1`] = `
<Modal
  onClick={[MockFunction]}
>
  <div
    className="dialog"
  >
    <div
      className="dialogTitle"
    >
      Request an invite
    </div>
    <Input
      errMsg={null}
      onEmitUpdates={[Function]}
      placeholder="Full Name"
    />
    <Input
      errMsg={null}
      onEmitUpdates={[Function]}
      placeholder="Email"
    />
    <Input
      errMsg={null}
      onEmitUpdates={[Function]}
      placeholder="Confirm Email"
      shouldEmitUpdates={true}
    />
  </div>
  <Button
    isDisabled={false}
    onClick={[Function]}
  >
    Send
  </Button>
</Modal>
`;

exports[`RequestInviteDialog input onChange validates should set correct error when onConfirmedEmailChanged is called: Error state for empty confirmed email 1`] = `
Object {
  "confirmEmailMsg": "Please input an Email address",
  "confirmedEmail": "",
  "email": "miles@ttt.com",
  "emailErrMsg": null,
  "fullName": "",
  "nameErrMsg": null,
}
`;

exports[`RequestInviteDialog input onChange validates should set correct error when onConfirmedEmailChanged is called: Error state for invalid email 1`] = `
Object {
  "confirmEmailMsg": "Please input a valid Email address",
  "confirmedEmail": "invalidCOnfimed",
  "email": "miles@ttt.com",
  "emailErrMsg": null,
  "fullName": "",
  "nameErrMsg": null,
}
`;

exports[`RequestInviteDialog input onChange validates should set correct error when onConfirmedEmailChanged is called: Error state for matched confirmed email 1`] = `
Object {
  "confirmEmailMsg": null,
  "confirmedEmail": "miles@ttt.com",
  "email": "miles@ttt.com",
  "emailErrMsg": null,
  "fullName": "",
  "nameErrMsg": null,
}
`;

exports[`RequestInviteDialog input onChange validates should set correct error when onConfirmedEmailChanged is called: Error state for not matched confirmed email 1`] = `
Object {
  "confirmEmailMsg": "Email address does not match",
  "confirmedEmail": "test@valid.com",
  "email": "miles@ttt.com",
  "emailErrMsg": null,
  "fullName": "",
  "nameErrMsg": null,
}
`;

exports[`RequestInviteDialog input onChange validates should set correct error when onEmailChanged is called: Error state for empty email 1`] = `
Object {
  "confirmEmailMsg": null,
  "confirmedEmail": "",
  "email": "",
  "emailErrMsg": "Please input an Email address",
  "fullName": "",
  "nameErrMsg": null,
}
`;

exports[`RequestInviteDialog input onChange validates should set correct error when onEmailChanged is called: Error state for invalid email 1`] = `
Object {
  "confirmEmailMsg": null,
  "confirmedEmail": "",
  "email": "invalid",
  "emailErrMsg": "Please input a valid Email address",
  "fullName": "",
  "nameErrMsg": null,
}
`;

exports[`RequestInviteDialog input onChange validates should set correct error when onEmailChanged is called: Error state for valid email 1`] = `
Object {
  "confirmEmailMsg": null,
  "confirmedEmail": "",
  "email": "test@valid.com",
  "emailErrMsg": null,
  "fullName": "",
  "nameErrMsg": null,
}
`;

exports[`RequestInviteDialog input onChange validates should set correct error when onNameChanged is called: Error state for empty name 1`] = `
Object {
  "confirmEmailMsg": null,
  "confirmedEmail": "",
  "email": "",
  "emailErrMsg": null,
  "fullName": "",
  "nameErrMsg": "You must input a name",
}
`;

exports[`RequestInviteDialog input onChange validates should set correct error when onNameChanged is called: Error state for short name 1`] = `
Object {
  "confirmEmailMsg": null,
  "confirmedEmail": "",
  "email": "",
  "emailErrMsg": null,
  "fullName": "ab",
  "nameErrMsg": "Full name must be at least 3 charactors long",
}
`;

exports[`RequestInviteDialog input onChange validates should set correct error when onNameChanged is called: Error state for valid name 1`] = `
Object {
  "confirmEmailMsg": null,
  "confirmedEmail": "",
  "email": "",
  "emailErrMsg": null,
  "fullName": "Good Name",
  "nameErrMsg": null,
}
`;

exports[`RequestInviteDialog snapshots should match snapshot in error status with dialogMessage 1`] = `
<Modal
  onClick={[MockFunction]}
>
  <div
    className="dialog"
  >
    <div
      className="dialogTitle"
    >
      Request an invite
    </div>
    <Input
      errMsg={null}
      onEmitUpdates={[Function]}
      placeholder="Full Name"
    />
    <Input
      errMsg={null}
      onEmitUpdates={[Function]}
      placeholder="Email"
    />
    <Input
      errMsg={null}
      onEmitUpdates={[Function]}
      placeholder="Confirm Email"
      shouldEmitUpdates={true}
    />
  </div>
  <Button
    isDisabled={true}
    onClick={[Function]}
  >
    Send
  </Button>
  <div
    className="serverErrorMessage"
  >
    EXAMPLE ERROR MESSAGE
  </div>
</Modal>
`;

exports[`RequestInviteDialog snapshots should match snapshot in inital status 1`] = `
<Modal
  onClick={[MockFunction]}
>
  <div
    className="dialog"
  >
    <div
      className="dialogTitle"
    >
      Request an invite
    </div>
    <Input
      errMsg={null}
      onEmitUpdates={[Function]}
      placeholder="Full Name"
    />
    <Input
      errMsg={null}
      onEmitUpdates={[Function]}
      placeholder="Email"
    />
    <Input
      errMsg={null}
      onEmitUpdates={[Function]}
      placeholder="Confirm Email"
      shouldEmitUpdates={true}
    />
  </div>
  <Button
    isDisabled={true}
    onClick={[Function]}
  >
    Send
  </Button>
</Modal>
`;

exports[`RequestInviteDialog snapshots should match snapshot in loading status 1`] = `
<Modal
  onClick={null}
>
  <div
    className="dialog"
  >
    <div
      className="dialogTitle"
    >
      Request an invite
    </div>
    <Input
      errMsg={null}
      onEmitUpdates={[Function]}
      placeholder="Full Name"
    />
    <Input
      errMsg={null}
      onEmitUpdates={[Function]}
      placeholder="Email"
    />
    <Input
      errMsg={null}
      onEmitUpdates={[Function]}
      placeholder="Confirm Email"
      shouldEmitUpdates={true}
    />
  </div>
  <Button
    isDisabled={true}
  >
    Sending, please wait...
  </Button>
</Modal>
`;

exports[`RequestInviteDialog snapshots should match snapshot in success status 1`] = `
<Modal
  onClick={[MockFunction]}
>
  <div
    className="dialog"
  >
    <div
      className="dialogTitle"
    >
      All Done!
    </div>
    <div
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      You will be one of the first to experience Broccoli & Co. when we launch.
    </div>
  </div>
  <Button
    onClick={[MockFunction]}
  >
    OK
  </Button>
</Modal>
`;
